lisp/lisp.go

```diff
@@ -77,8 +84,13 @@ const (
 	//  	[1] a list containing row-major ordered array values
 	LArray
 	// LNative values store a Go value in the LVal.Native field and can be used
-	// by builtin functions store references to values of any type.
+	// by builtin functions to store values of any type.
 	LNative
+	// LTaggedVal is a user-defined type that uses the following fields in an
+	// LVal:
+	// 		LVal.Str      The user-defined type name
+	// 		LVal.Cells[0] The user-data for the typed-value
+	LTaggedVal
```

Comment:
I'm not crazy about the name "tagged-value". But I wasn't sure what else to use and I think I got "tagged-value" from SICP. I didn't really want to call it an "instance".

I was also thinking about the name "user-data" -- data with a user-defined type. I think this name sounds better but in the future some "user data" may be part of the core language (arrays, maps, ...)

Reply:
I think the term "tagged-value" is fine along with `LTaggedVal`. But after mulling it over I think I dislike the lisp function `(tagged-value v)` enough to change it -- it's a very confusing function name. I think the data in a tagged-value should be called "user-data", data managed by the "user" (program) that defined the type. So you'd write something like

```lisp
(if (tagged-value? v)
  (do-something (user-data v))
  (error 'type-error "argument is not a tagged-value")
```

And for the function `new` you would say "new takes a typedef and returns a tagged-value with the given type and user-data returned by the type constructor after being invoked with new's remaining arguments".